### PR TITLE
fix checkout metadata fetches to not create checkout metadata objects

### PR DESCRIPTION
**Pull request by jakubkuc**
I want to merge this change because it prevents fetches of checkout metadata to create checkout metadata objects in db. Otherwise it can cause a thread race during shipping method and payment method fetching during `checkoutCreate`.

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/checkout/utils.py [hunk 1]
```diff
@@ -969,3 +969,10 @@ def get_or_create_checkout_metadata(checkout: "Checkout"):
         return checkout.metadata_storage
     else:
         return CheckoutMetadata.objects.create(checkout=checkout)
+
+
+def get_checkout_metadata(checkout: "Checkout"):
```
Comment (korycins) [inline saleor/checkout/utils.py:974]: # 

In my opinion, we should double-check all occurrences of `get_or_create_checkout_metadata` and replace them with `get_checkout_metadata`. If the issue has already occurred once, it's only a matter of time before `get_or_create_checkout_metadata` causes the same issue again. Using `get_or_create_checkout_metadata` in one thread can create problems with saving metadata in a second thread. If your checkout doesn't have `metadata_storage`, you create a dummy object and maybe call `.save()` on it. If another thread calls `get_or_create_checkout_metadata` in the meantime, we will get an exception that the object already exists..

Comment (jakubkuc) [reply saleor/checkout/utils.py:974]: You are right. I've missed some places that should be modified. Fixed it.

**Review by IKarbowiak — APPROVED** [review]

**Review by korycins — APPROVED** [review]

File: saleor/checkout/utils.py [hunk 2]
```diff
@@ -969,3 +969,10 @@ def get_or_create_checkout_metadata(checkout: "Checkout"):
...
+    if hasattr(checkout, "metadata_storage"):
+        return checkout.metadata_storage
+    else:
+        return CheckoutMetadata()
```
Comment (maarcingebala) [inline saleor/checkout/utils.py:978]: This is risky - this instance would not be associated with the given checkout instance, saving that into DB would probably fail.

File: saleor/checkout/utils.py (hunk 1 above)
Comment (maarcingebala) [inline saleor/checkout/utils.py:974]: Can we add unit tests for this function? They should check both scenarios - saving metadata to a checkout where storage already exists and to a checkout without it.

File: saleor/graphql/checkout/tests/mutations/test_checkout_complete.py [hunk 3]
```diff
@@ -484,6 +484,84 @@ def test_checkout_complete_with_metadata_updates_existing_keys(
...
+
+    # when
+    response = user_api_client.post_graphql(MUTATION_CHECKOUT_COMPLETE, variables)
+    content = get_graphql_content(response)
+    data = content["data"]["checkoutComplete"]
+
+    # then
+    assert not data["errors"]
+    assert Order.objects.count() == orders_count + 1
+    order = Order.objects.first()
```
Comment (tomaszszymanski129) [inline saleor/graphql/checkout/tests/mutations/test_checkout_complete.py:548]: @IKarbowiak Can we trust it to always be the correct one? (same for other test)

**Review by tomaszszymanski129 — APPROVED** [review]

File: saleor/graphql/checkout/tests/mutations/test_checkout_complete.py (hunk 3 above)
Comment (tomaszszymanski129) [reply saleor/graphql/checkout/tests/mutations/test_checkout_complete.py:548]: Agreed that refactor will be handled in separate ticket as it involves also other tests.

Comment (IKarbowiak) [reply saleor/graphql/checkout/tests/mutations/test_checkout_complete.py:548]: Issue created: https://github.com/saleor/saleor/issues/12728

File: saleor/checkout/utils.py (hunk 1 above)
Comment (IKarbowiak) [reply saleor/checkout/utils.py:974]: Done

**Pull request merged**